File: app/src/main/java/com/islam/basepropject/project_base/base/adapters/BaseAdapter.java
package com.islam.basepropject.project_base.base.adapters;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.islam.basepropject.project_base.views.MyRecyclerView;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseAdapter<T, VH extends BaseViewHolder<T>> extends RecyclerView.Adapter<VH> {

    protected List<T> list;

    public BaseAdapter() {
    }

    public BaseAdapter(List<T> list) {
        this.list = list;
    }

    public void registerAdapterDataObservertion(MyRecyclerView recyclerView) {
        registerAdapterDataObserver(new AdapterDataObservation(recyclerView));
    }

    @NonNull
    @Override
    public abstract VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType);

    @Override
    public final void onBindViewHolder(@NonNull VH holder, int position) {
        holder.onBind(list.get(position));
    }

    @Override
    public int getItemCount() {
        if (list != null)
            return list.size();
        return 0;
    }

    public void setData(List<T> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    public void addItem(T item) {
        initList();
        list.add(item);
        notifyItemInserted(getItemCount() - 1);
    }

    public void addItem(T item, int position) {
        initList();
        list.add(position, item);
        notifyItemInserted(position);
    }

    public void updateItem(T item) {
        int index = list.indexOf(item);
        if (index != -1)
            updateItem(item, index);
    }

    public void updateItem(T item, int position) {
        initList();
        list.set(position, item);
        notifyItemChanged(position);
    }

    public void removeItem(T item) {
        int index = list.indexOf(item);
        if (index != -1)
            removeItem(index);
    }

    public void removeItem(int position) {
        initList();
        list.remove(position);
        notifyItemRemoved(position);
    }

    public void clear() {
        initList();
        list.clear();
        notifyDataSetChanged();
    }

    private void initList() {
        if (list == null)
            list = new ArrayList<>();
    }
}

File: app/src/main/java/com/islam/basepropject/project_base/base/adapters/BaseAdapter.kt
package com.islam.basepropject.project_base.base.adapters

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

import com.islam.basepropject.project_base.views.MyRecyclerView

import java.util.ArrayList

abstract class BaseAdapter<T, VH : BaseViewHolder<T>> : RecyclerView.Adapter<VH> {

    protected var list: MutableList<T>? = null

    constructor() {}

    constructor(list: MutableList<T>) {
        this.list = list
    }

    fun registerAdapterDataObservertion(recyclerView: MyRecyclerView) {
        registerAdapterDataObserver(AdapterDataObservation(recyclerView))
    }

    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH

    override fun onBindViewHolder(holder: VH, position: Int) {
        holder.onBind(list!![position])
    }

    override fun getItemCount(): Int {
        return if (list != null) list!!.size else 0
    }

    fun setData(list: MutableList<T>) {
        this.list = list
        notifyDataSetChanged()
    }

    fun addItem(item: T) {
        initList()
        list!!.add(item)
        notifyItemInserted(itemCount - 1)
    }

    fun addItem(item: T, position: Int) {
        initList()
        list!!.add(position, item)
        notifyItemInserted(position)
    }

    fun updateItem(item: T) {
        val index = list!!.indexOf(item)
        if (index != -1)
            updateItem(item, index)
    }

    fun updateItem(item: T, position: Int) {
        initList()
        list!![position] = item
        notifyItemChanged(position)
    }

    fun removeItem(item: T) {
        val index = list!!.indexOf(item)
        if (index != -1)
            removeItem(index)
    }

    fun removeItem(position: Int) {
        initList()
        list!!.removeAt(position)
        notifyItemRemoved(position)
    }

    fun clear() {
        initList()
        list!!.clear()
        notifyDataSetChanged()
    }

    private fun initList() {
        if (list == null)
            list = ArrayList()
    }
}